rfd/XXXX-upgrade-system.md

```diff
@@ -0,0 +1,1658 @@
+a combination of both persistent backend records, and "live" control stream status.
+
+Given the above, the reconciliation loop is a bit more complex, but still falls into three distinct phases:
+
+1. Setup: load cluster-level upgrade system configuration, active `version-directive`, churn/fault stats, etc.
+
+2. Reconciliation: Match servers to target and installer, and categorize them by their current
+install eligibility given recent install attempts, control stream status, etc.
+
+3. Application: Determine the number of eligible servers that will actually be slated for install
+given current target rollout rate, update their backend states with a summary of the install attempt
+that is about to be made (skipping servers which had their installation status concurrently updated),
+and pass them off to installer-specific logic.
+
+As much as possible, we want the real "decision making" power to rest with the `version-controller` rather
+than the version reconciliation loop. That being said, the version reconciliation loop will have some
+internal rules that it evaluates to make sure that directives, as applied to the current server inventory,
+do not result in any invalid state-transitions (e.g. it will refuse to change the target arch for a given
+server, or skip a major version).
+
+
+### Version Directives
+
+#### The Version Directive Resource
+
+The `version-directive` resource is the heart of the upgrade system. It is a static resource that describes
+the current desired state of the cluster and how to get to that state. This is achieved through a series of
+matchers which are used to pair servers with installation targets and installers. At its core, a `version-directive`
+can be thought of as a function of the form `f(server) -> optional(target,installer)`.
+
+Installation targets are arbitrary attribute mappings that must *at least* contain `version`, but may contain
+any additional information as well. Certain metadata is understood by teleport (e.g. `fips:yes|no`,
+`arch:amd64|arm64|...`), but additional metadata (e.g. `sha256sum:12345...`) is simply passed through
+to the installer.  The target to be used for a given server is the first target that
+*is not incompatible* (i.e. no attempt to find the "most compatible" target is made).  A target is
+incompatible with a server if that server's version cannot safely upgrade/downgrade to that target version,
+*or* if the target specifies a build attribute that differs from a build attribute of the current
+installation (e.g. `fips:yes` when current build is `fips:no`). We don't require that all build attributes
+are present since not all systems require knowledge of said attributes.  It is the responsibility of
+an installer to fail if it is not provided with sufficient attributes to perform the installation
+safely (e.g. the `tuf` installer would fail if the target passed to it did not contain the expected
+length and hash data).
```

Comment:
Would said installer still get matched?

Reply:
Checking expected target attributes is definitely something we *could* keep track of for matching.  My current feeling is that it might be better not to do this.  I'm wary of making the system too clever, as I want to keep it easy to review and reason about. Installer selection will definitely need to take node version into account since there will be an intermediate time period where folks may have some nodes that support `tuf` installers, but some that only support `script`. Additionally, one of the open questions is about whether or not to support nodes explicitly opting into a limited subset of the available installers (the most obvious choice being to explicitly only allow the use of the `tuf` installer, since it is likely to be the most secure once it is stabilized). If we did something like that, we'd definitely want it to be part of the matching logic as well.